### DISCUSSION TITLE
Recommend way to implement login for frontend?

### DISCUSSION THREAD

**Discussion by CrescentByte** [pimcore/pimcore#14914] [Q&A / Support]
Hi all. I am currently working on a pimcore based website and would like to implement a way for users to login to the frontend specifically, as opposed to the provided administration backend. I have scoured the net far and wide, but I cannot seem to find any tutorials, documentation or advice on how to implement this. I am hoping that there is some way to reuse the existing implementation provided by the admin side, but redirect to the frontend instead of the backend once the user logs in. And obviously most users are not going to be an administrator or require access to the administration side of the site. Presumably, reusing the existing implementation in some way and redirecting to the frontend would be the recommended way of doing this but please correct me if I am wrong here. I was thinking of setting up a custom twig template with a login form that utilizes the same logic/handling as the login for the admin side and then redirects to the home page after the user successfully logs in, but am unsure of how to point a custom twig based login form to the admin login handling and then redirecting a successful login attempt to the home(frontend) page. Any help, advice or useful information regarding this would be amazing, thanks 👍

**Comment by passioneight** [comment]
Symfony's documentation about [security](https://symfony.com/doc/current/security.html) should get you started. Also, you may want to check out [Pimcore's CMF](https://pimcore.com/docs/customer-management-framework/current/index.html).

**Reply by aarongerig** [reply]
There's also this bundle, which includes authentication functionality: [Pimcore Members](https://github.com/dachcom-digital/pimcore-members/blob/3.x/UPGRADE.md)

**Reply by CrescentByte** [reply]
> Symfony's documentation about [security](https://symfony.com/doc/current/security.html) should get you started. Also, you may want to check out [Pimcore's CMF](https://pimcore.com/docs/customer-management-framework/current/index.html).

The security doc seems to have a partial example that is incompatible with pimcore 😔 It feels like all this documentation is just bits and pieces and I don't understand how to actually adapt this into a functioning login system for the frontend of my pimcore site. Even googling "Pimcore frontend user login" doesn't return anything helpful for what I am trying to achieve, [see for yourself](https://www.google.com/search?q=pimcore+frontend+user+login). Even the devs of pimcore themselves have made no commentary, recommendation or announcement on what the correct way to implement a frontend login is, despite it being an obviously critical and fundamental building block to any pimcore site that's frontend focused. I also tried checking the CMF link but it seems to be a whole thing, complete with ecommerce stuff in it, whereas I just want a simple user login system (with registration obviously) for the frontend. I'm running pimcore v10.5.14 build 33d8b3b17ebb77c733c304da5fef9954279c089b, on a docker image (pimcore skeleton), on a manjaro device. I saw this: [Authenticate against Pimcore Objects](https://pimcore.com/docs/pimcore/current/Development_Documentation/Development_Tools_and_Details/Security_Authentication/Authenticate_Pimcore_Objects.html) but it too seems to be an incomplete and not working example, I tried hooking up the user wrapper class to the default admin user class at first, and it said I couldn't extend from it because it was a final class, and then I tried creating a custom data object class that had the same fields shown in the first screenshot on the page, but I couldn't find the 'roles' field anywhere. I checked all the possible options in the context menu to add to the object and none of the options is named 'roles' and none of the icons match the one displayed in the screenshot. It just seems I can't find an example of what I want, Is there not a step by step tutorial somewhere out there, that explains how to make a user login + registration system for the frontend that explains exactly what files need to be edited and what classes need to be create in order for it to work, and what yaml configs need to be edited? I've been trying to get this working all week and so far I have nothing to show for it, it's starting to stress me out because I thought this would have been explained somewhere and fairly straight forward, but there seems to only be bits and pieces of outdated documentation and it's impossible to see the full picture here. Bassically I want:

- Either a custom user class (Pimcore data object) or the standard pimcore admin user
  
- Controller class that interacts with the user class (either a custom one or pimcore admin one) that defines (presumably) 3 routes, login, logout and register.
  
- Login form twig template.
  
- Register form twig template.
  
- Whatever config files need to be edited(cant't find a tutorial on how to make a custom user provider for pimcore) and whatever changes to the firewall config need to be made to get this all to work.
  

> There's also this bundle, which includes authentication functionality: [Pimcore Members](https://github.com/dachcom-digital/pimcore-members/blob/3.x/UPGRADE.md)

How do i use this? Can't find instructions on how to use properly 😅 how do I use this to make a login form and a registration form to create new users on the frontend of my site?

**Reply by passioneight** [reply]
Could you confirm if you have checked the correct version of Symfony's documentation?

Also, I noticed that [the link you shared](https://pimcore.com/docs/pimcore/current/Development_Documentation/Development_Tools_and_Details/Security_Authentication/Authenticate_Pimcore_Objects.html) mentions the demo [here](https://github.com/pimcore/demo/blob/41e241296e6a6c0dcdf6b14120ab416dabc29452/config/packages/security.yaml#L58). I recommend to set it up and play around - maybe even debug - to see how this is working.

By the way, there is a preview of the upcoming documentation update available in the latest newsletter: https://pimcore.com/docs/platform/Customer_Management_Framework. Maybe this helps as well.

**Reply by CrescentByte** [reply]
> Could you confirm if you have checked the correct version of Symfony's documentation?

I checked 5.4 because that's the version my pimcore skeleton docker image is running, according to the debug bar at the bottom of the page on my site.

The demo is a huge project that extends well beyond a simple login and register setup, a bit overwhelming for me to be honest. That's why i wanted to start out small, with something simple(or so I thought). Thanks to chat gpt, I managed to get a login form working and 90% of the way with a registration form, only thing left is I need to find a way to create a new instance of the pimcore user object programatically, basically what you do in the admin ui to create a new non-admin user, but in code instead of using the ui to do it. trying to find out how to do that has been a bit of a pain and google isn't helping 🙃 I have tried creating a new user object and then using the $user->save() method on the new pimcore user object to save it to the database once the fields from the form are set and it is saving to the database, but It's still not showing up in the admin ui user list, and the password is not hashed either(checked security.yml and didn't see anything in the way of password hashing being setup by the admin bundle so not sure how the admin bundle is doing it), and trying to login to the main firewall as the new user does not work, probably related to the first 2 issues most likely. So, my question now is what is the correct way to create a new pimcore user object in code correctly, so that the password is hashed and the account can be logged into? Tried checking the documentation and googling things like "how to create a pimcore user programatically" and "how to make a new pimcore user in code", even asked chat gpt but found nothing that works so far, if someone would be kind enough to explain to me how to do this properly it would be greatly appreciated!

**Reply by CrescentByte** [reply]
Okay so I managed to find an example of how to create a user in code here https://pimcore.com/docs/pimcore/current/Development_Documentation/Development_Tools_and_Details/Extending_a_Backend_User.html this uses the user::create() method and includes a few extra parameters that I wasn't including before like the parent id, has_credentials and active attributes. This way, I managed to create the user and it showed up in the admin ui, however the password still wasn't hashed, so I tried manually hashing it with the password_hash() function and specifying bcrypt as the algorithm as I believe that's what pimcore uses for password hashing, but i still cannot login as the new user, even though I dumped the output of password_veryify() with the plaintext password and the generated hash, and it returned true. So not sure why it isn't working, as if there is some other step i am missing. I noticed that if I changed/set the password of the new user in the admin ui, i can login as the new user, so I know that the password hashing is the culprit, the final obstacle in getting this working!!
